native/src/accelerated_bn/bn254.rs

```diff
@@ -0,0 +1,183 @@
+        g1: impl Iterator<Item = <Bn254 as Pairing>::G1Prepared>,
+        g2: impl Iterator<Item = <Bn254 as Pairing>::G2Prepared>,
+    ) -> Result<<Bn254 as Pairing>::TargetField, ()> {
+        let g1 = utils::encode(g1.collect::<Vec<_>>());
+        let g2 = utils::encode(g2.collect::<Vec<_>>());
+        let res = host_calls::bn254_multi_miller_loop(g1, g2).unwrap_or_default();
+        utils::decode(res).map_err(|_| ())
+    }
+
+    fn bn254_final_exponentiation(
+        target: <Bn254 as Pairing>::TargetField,
+    ) -> Result<<Bn254 as Pairing>::TargetField, ()> {
+        let target = utils::encode(target);
+        let res = host_calls::bn254_final_exponentiation(target).unwrap_or_default();
+        utils::decode(res).map_err(|_| ())
+    }
+
+    fn bn254_msm_g1(
+        bases: &[G1Affine],
+        scalars: &[<G1Config as CurveConfig>::ScalarField],
+    ) -> Result<G1Projective, ()> {
+        let bases = utils::encode(bases);
+        let scalars = utils::encode(scalars);
+        let res = host_calls::bn254_msm_g1(bases, scalars).unwrap_or_default();
+        utils::decode_proj_sw(res).map_err(|_| ())
+    }
+
+    fn bn254_msm_g2(
+        bases: &[G2Affine],
+        scalars: &[<G2Config as CurveConfig>::ScalarField],
+    ) -> Result<G2Projective, ()> {
+        let bases = utils::encode(bases);
+        let scalars = utils::encode(scalars);
+        let res = host_calls::bn254_msm_g2(bases, scalars).unwrap_or_default();
+        utils::decode_proj_sw(res).map_err(|_| ())
+    }
+
+    fn bn254_mul_projective_g1(base: &G1Projective, scalar: &[u64]) -> Result<G1Projective, ()> {
+        let base = utils::encode_proj_sw(base);
+        let scalar = utils::encode(scalar);
+        let res = host_calls::bn254_mul_projective_g1(base, scalar).unwrap_or_default();
+        utils::decode_proj_sw(res).map_err(|_| ())
+    }
+
+    fn bn254_mul_projective_g2(base: &G2Projective, scalar: &[u64]) -> Result<G2Projective, ()> {
+        let base = utils::encode_proj_sw(base);
+        let scalar = utils::encode(scalar);
+        let res = host_calls::bn254_mul_projective_g2(base, scalar).unwrap_or_default();
+        utils::decode_proj_sw(res).map_err(|_| ())
+    }
+}
+
+/// Interfaces for working with *Arkworks* *BN254* elliptic curve related types
+/// from within the runtime.
+///
+/// All types are (de-)serialized through the wrapper types from the `ark-scale` trait,
+/// with `ark_scale::{ArkScale, ArkScaleProjective}`.
+///
+/// `ArkScale`'s `Usage` generic parameter is expected to be set to "not-validated"
+/// and "not-compressed".
+#[runtime_interface]
+pub trait HostCalls {
+    /// Pairing multi Miller loop for *BN254*.
+    ///
+    /// - Receives encoded:
+    ///   - `a`: `ArkScale<Vec<G1Affine>>`.
+    ///   - `b`: `ArkScale<Vec<G2Affine>>`.
+    /// - Returns encoded:  `ArkScale<Bn254::TargetField>`.
+    fn bn254_multi_miller_loop(a: Vec<u8>, b: Vec<u8>) -> Result<Vec<u8>, BnCryptoError> {
```

Comment:
I would suggest returning a `()` error, to be coherent with `sp-crypto-ec-utils`. You can silence clippy warning byadding a `#[allow(clippy::result_unit_err)]` attribute to the trait. The same for all the other functions of the trait.